scheduler/context.go

```diff
@@ -90,6 +89,40 @@ type PortCollisionEvent struct {
 	NetIndex *structs.NetworkIndex
 }
 
+func (ev *PortCollisionEvent) Copy() *PortCollisionEvent {
+	if ev == nil {
+		return nil
+	}
+	c := new(PortCollisionEvent)
+	*c = *ev
+	c.Node = ev.Node.Copy()
+	if len(ev.Allocations) > 0 {
+		for i, a := range ev.Allocations {
+			c.Allocations[i] = a.Copy()
+		}
+
+	}
+	c.NetIndex = ev.NetIndex.Copy()
+	return c
+}
+
+func (ev *PortCollisionEvent) Sanitize() *PortCollisionEvent {
+	if ev == nil {
+		return nil
+	}
+	clean := ev.Copy()
+
+	clean.Node = ev.Node.Sanitize()
+	clean.Node.Meta = make(map[string]string, 0)
+
+	for i, alloc := range ev.Allocations {
+		clean.Allocations[i] = alloc.Copy()
```

Comment:
This should skip a lot of work:

```suggestion
		clean.Allocations[i] = alloc.CopySkipJob()
```

Note you still need to explicitly nil out the Job. This method should be named `CopyShallowJob()` or something because it shallow copies the job (whereas plain old Copy deep copies it).

Reply:
TIL, thanks!